google-cloud-core/src/main/java/com/google/cloud/BaseServiceException.java

```diff
@@ -140,20 +141,36 @@ public BaseServiceException(IOException exception, boolean idempotent) {
     String location = null;
     String debugInfo = null;
     Boolean retryable = null;
-    if (exception instanceof GoogleJsonResponseException) {
-      GoogleJsonError jsonError = ((GoogleJsonResponseException) exception).getDetails();
-      if (jsonError != null) {
-        Error error = new Error(jsonError.getCode(), reason(jsonError));
-        code = error.code;
-        reason = error.reason;
-        retryable = isRetryable(idempotent, error);
-        if (reason != null) {
-          GoogleJsonError.ErrorInfo errorInfo = jsonError.getErrors().get(0);
-          location = errorInfo.getLocation();
-          debugInfo = (String) errorInfo.get("debugInfo");
+    if (exception instanceof HttpResponseException) {
+      if (exception instanceof GoogleJsonResponseException) {
+        GoogleJsonError jsonError = ((GoogleJsonResponseException) exception).getDetails();
+        if (jsonError != null) {
+          Error error = new Error(jsonError.getCode(), reason(jsonError));
+          code = error.code;
+          reason = error.reason;
+          retryable = isRetryable(idempotent, error);
+          if (reason != null) {
+            GoogleJsonError.ErrorInfo errorInfo = jsonError.getErrors().get(0);
+            location = errorInfo.getLocation();
+            debugInfo = (String) errorInfo.get("debugInfo");
+          }
+        } else {
+          code = ((GoogleJsonResponseException) exception).getStatusCode();
         }
       } else {
-        code = ((GoogleJsonResponseException) exception).getStatusCode();
+        // In cases where an exception is an instance of HttpResponseException but not
+        // an instance of GoogleJsonResponseException, check the status code to determine whether it's retryable
+        code = ((HttpResponseException) exception).getStatusCode();
+        Set<Error> retryableErrors = getRetryableErrors();
+        if (retryableErrors != null && !retryableErrors.isEmpty()) {
+          for (Error e : retryableErrors) {
+            if (e.getCode() != null && e.getCode().equals(code)) {
+              Error error = new Error(code, null);
```

Comment:
The body of the loop does not depend on the loop itself (the "retriable" is assigned based on value of "code" and "idempotent", both of which are known outside of the loop), potentially a bug. 

Looks like the tests pass, because the actual "isRetryable(boolean idempotent, Error error)" call already iterates over the getRetriableErrors() in its implementation. So, probably the code can be simplified to the single direct call to that method:
retriable = isRetriable(idempotent, new Error(((HttpResponseException) exception).getStatusCode(), null));